File: scripts/weather_bundle.py
"""Build a deterministic weather publication bundle after retained-source replay."""
import hashlib
import json
from pathlib import Path
from weather_history import build as replay_history
from venue_evidence import validate_venue
from refresh_weather import parse_time

ROOT = Path(__file__).resolve().parents[1]


def encode(value):
    return json.dumps(value, sort_keys=True, separators=(',', ':'), allow_nan=False).encode()


def fingerprint(value):
    return hashlib.sha256(encode(value)).hexdigest()


def build(root=ROOT):
    def read(name): return json.loads((root / 'data' / name).read_bytes())
    weather, history, site = read('weather.json'), read('weather-history.json'), read('site.json')
    if replay_history(root) != history:
        raise ValueError('Weather history differs from retained-source replay')
    started, generated = parse_time(weather['collectionStartedAt']), parse_time(weather['generatedAt'])
    if started > generated:
        raise ValueError('Invalid weather collection interval')
    venues, extra = read('weather-venues.json'), read('weather-osm-venues.json')
    if any(venues[k].get('status') != 'unavailable' for k in venues.keys() & extra.keys()):
        raise ValueError('Conflicting venue registry')
    venues.update(extra)
    for venue in venues.values(): validate_venue(venue)
    games = {game['id']: game for game in site['games']}
    if len(games) != len(site['games']) or set(games) != set(weather['games']):
        raise ValueError('Weather game context scope differs')
    ledger = read('weather-ledger.json')
    retained = {record['hash']: record for record in ledger}
    contexts = {}
    for identity, record in weather['games'].items():
        game = games[identity]
        if record['gameId'] != identity or any(record.get(k) != game.get(k) for k in ('kickoff', 'venue')):
            raise ValueError('Weather game context changed')
        contexts[identity] = {k: game.get(k) for k in ('id', 'season', 'kickoff', 'venue', 'neutral')}
        if record['status'] == 'available':
            venue = venues.get(game['venue'])
            if game.get('neutral') or not venue or record.get('locationEvidence') != venue:
                raise ValueError('Weather venue context changed')
            if 'validFromSeason' in venue and game['season'] < venue['validFromSeason']:
                raise ValueError('Weather venue season changed')
            if not started <= parse_time(record['retrievedAt']) <= generated:
                raise ValueError('Observation outside weather collection interval')
            digest = fingerprint(record)
            if retained.get(digest) != {**record, 'hash': digest}:
                raise ValueError('Current weather observation is not retained')
    payload = {'schemaVersion': 1, 'weather': weather, 'history': history, 'contexts': contexts,
               'venueRegistryHash': fingerprint(venues)}
    sources = sorted({record[key] for record in ledger for key in ('pointHash', 'sourceHash')})
    manifest = {'schemaVersion': 1, 'validatorVersion': 'weather-bundle-v1',
                'collectionStartedAt': weather['collectionStartedAt'], 'generatedAt': weather['generatedAt'],
                'payloadSha256': fingerprint(payload), 'payloadBytes': len(encode(payload)),
                'venueRegistryHash': payload['venueRegistryHash'], 'sourceHashes': sources,
                'ledgerSha256': fingerprint(ledger)}
    return {'manifest': manifest, 'payload': payload}


if __name__ == '__main__':
    result = build()
    output = ROOT / 'release-recovery/weather-bundle.json'
    output.parent.mkdir(exist_ok=True)
    output.write_bytes(encode(result) + b'\n')
    print(json.dumps(result['manifest'], indent=2))

File: src/lib/weather-bundle.ts
import {createHash} from 'node:crypto';
import {isDeepStrictEqual} from 'node:util';
import type {WeatherRecord} from './weather';
import {weatherHistoryForGame,type WeatherHistory} from './weather-history';

const digest=(body:Buffer)=>createHash('sha256').update(body).digest('hex');
const hash=(v:unknown)=>typeof v==='string'&&/^[a-f0-9]{64}$/.test(v);
const time=(v:unknown)=>typeof v==='string'&&/(Z|[+-]\d\d:\d\d)$/.test(v)&&Number.isFinite(Date.parse(v));
const object=(v:unknown)=>v!==null&&typeof v==='object'&&!Array.isArray(v);
type Context={id:string;season:number;kickoff:string|null;venue:string;neutral:boolean};
function sameVenue(a:unknown,b:unknown,key=''):boolean{
  if(typeof a==='number'&&typeof b==='number'&&['latitude','longitude','minlat','minlon','maxlat','maxlon'].includes(key)){
    // Bundlers can shorten JSON numeric literals. This is below 0.02 mm on Earth.
    return Number.isFinite(a)&&Number.isFinite(b)&&Math.abs(a-b)<=1e-10;
  }
  if(object(a)&&object(b)){
    const left=a as Record<string,unknown>,right=b as Record<string,unknown>;
    return Object.keys(left).length===Object.keys(right).length&&Object.keys(left).every(k=>Object.hasOwn(right,k)&&sameVenue(left[k],right[k],k));
  }
  return isDeepStrictEqual(a,b);
}
export type WeatherBundle={
  weather:{collectionStartedAt:string;generatedAt:string;games:Record<string,WeatherRecord>};
  history:WeatherHistory;contexts:Record<string,Context>;venueRegistryHash:string;
  sourceHashes:string[];ledgerSha256:string;
};

export function assertWeatherContinuity(previous:WeatherBundle,next:WeatherBundle){
  if(Date.parse(next.weather.generatedAt)<Date.parse(previous.weather.generatedAt))throw Error('Weather collection regressed');
  const rows=new Map(next.history.records.map(row=>[row.hash,row]));
  if(previous.history.records.some(row=>!isDeepStrictEqual(rows.get(row.hash),row)))throw Error('Retained weather history was removed or changed');
  const sources=new Set(next.sourceHashes);
  if(previous.sourceHashes.some(hash=>!sources.has(hash)))throw Error('Retained weather source was removed');
}

export function boundWeatherRecord(bundle:WeatherBundle,game:Context,venues:Record<string,unknown>):WeatherRecord|undefined{
  const context=bundle.contexts[game.id],record=bundle.weather.games[game.id];
  if(!context||!record||!['id','season','kickoff','venue','neutral'].every(key=>context[key as keyof Context]===game[key as keyof Context]))return undefined;
  if(record.status==='available'){
    const approved=Object.hasOwn(venues,game.venue)?venues[game.venue]:null;
    if(game.neutral||!approved||!sameVenue(record.locationEvidence,approved))return undefined;
  }
  return record;
}

/** Verify transport/bundle consistency; raw NWS replay remains the trusted publisher's job. */
export function decodeWeatherBundle(raw:Buffer,now=Date.now()):WeatherBundle{
  if(raw.length>2_000_000)throw Error('Weather bundle too large');
  const envelope=JSON.parse(raw.toString()),meta=envelope?.manifest;
  if(!object(meta)||meta.schemaVersion!==1||meta.validatorVersion!=='weather-bundle-v1'||typeof envelope.payloadJson!=='string')throw Error('Unsupported weather bundle');
  const bytes=Buffer.from(envelope.payloadJson);
  if(bytes.length!==meta.payloadBytes||digest(bytes)!==meta.payloadSha256)throw Error('Weather payload integrity failure');
  if(!time(meta.collectionStartedAt)||!time(meta.generatedAt)||!Number.isFinite(now)||Date.parse(meta.collectionStartedAt)>Date.parse(meta.generatedAt)||Date.parse(meta.generatedAt)>now)throw Error('Invalid weather bundle time');
  if(!hash(meta.venueRegistryHash)||!hash(meta.ledgerSha256)||!Array.isArray(meta.sourceHashes)||meta.sourceHashes.length>4096||meta.sourceHashes.some((h:unknown)=>!hash(h))||new Set(meta.sourceHashes).size!==meta.sourceHashes.length)throw Error('Invalid weather evidence references');
  const payload=JSON.parse(envelope.payloadJson),weather=payload?.weather,history=payload?.history,contexts=payload?.contexts;
  if(payload?.schemaVersion!==1||payload.venueRegistryHash!==meta.venueRegistryHash||!object(weather)||weather.collectionStartedAt!==meta.collectionStartedAt||weather.generatedAt!==meta.generatedAt||!object(weather.games)||!object(contexts)||!object(history)||history.schemaVersion!==1||!Array.isArray(history.records)||history.records.length>20000)throw Error('Invalid weather payload schema');
  const ids=Object.keys(weather.games);
  if(ids.length>1000||ids.length!==Object.keys(contexts).length)throw Error('Invalid weather context scope');
  for(const row of history.records){
    if(!object(row)||!hash(row.hash)||!hash(row.locationHash)||!hash(row.sourceHash)||!meta.sourceHashes.includes(row.sourceHash)||typeof row.gameId!=='string'||typeof row.venue!=='string'||!time(row.kickoff)||!time(row.issuedAt)||!time(row.retrievedAt)||Date.parse(row.retrievedAt)>Date.parse(meta.generatedAt))throw Error('Invalid weather history record');
  }
  if(new Set(history.records.map((r:{hash:string})=>r.hash)).size!==history.records.length)throw Error('Duplicate weather history');
  for(const id of ids){
    const record=weather.games[id],context=contexts[id];
    if(!object(record)||!object(context)||context.id!==id||!Number.isInteger(context.season)||typeof context.neutral!=='boolean'||typeof context.venue!=='string'||(context.kickoff!==null&&!time(context.kickoff))||record.gameId!==id||record.venue!==context.venue||record.kickoff!==context.kickoff||!['available','unavailable'].includes(record.status))throw Error('Invalid weather game context');
    if(record.status==='available'){
      if(context.neutral||!hash(record.sourceHash)||!hash(record.pointHash)||!meta.sourceHashes.includes(record.sourceHash)||!meta.sourceHashes.includes(record.pointHash)||!time(record.retrievedAt)||Date.parse(record.retrievedAt)<Date.parse(meta.collectionStartedAt)||Date.parse(record.retrievedAt)>Date.parse(meta.generatedAt)||!weatherHistoryForGame(history,record,context))throw Error('Current weather lacks matching retained history');
    }
  }
  return {weather,history,contexts,venueRegistryHash:meta.venueRegistryHash,sourceHashes:meta.sourceHashes,ledgerSha256:meta.ledgerSha256};
}
